lib/Runtime/Library/PropertyString.h

```diff
@@ -6,58 +6,86 @@
+        Var propertyValue,
+        ScriptContext *const requestContext,
+        const PropertyOperationFlags propertyOperationFlags,
+        PropertyValueInfo *const propertyValueInfo);
+
+    bool TryGetPropertyFromCache(
+        Var const instance,
+        RecyclableObject *const object,
+        Var *const propertyValue,
+        ScriptContext *const requestContext,
+        PropertyValueInfo *const propertyValueInfo);
+
+    static PropertyString* New(StaticType* type, const Js::PropertyRecord* propertyRecord, Recycler *recycler);
+
+    virtual void const * GetOriginalStringReference() override;
+    virtual RecyclableObject * CloneToScriptContext(ScriptContext* requestContext) override;
+
+    static uint32 GetOffsetOfLdElemInlineCache() { return offsetof(PropertyString, ldElemInlineCache); }
+    static uint32 GetOffsetOfStElemInlineCache() { return offsetof(PropertyString, stElemInlineCache); }
+    static uint32 GetOffsetOfHitRate() { return offsetof(PropertyString, hitRate); }
+    static bool Is(Var var);
+    static bool Is(RecyclableObject * var);
+
+    template <typename T> static PropertyString* TryFromVar(T var);
 
 #if ENABLE_DEBUG_CONFIG_OPTIONS
-        void DumpCache(bool ldElemCache)
+    void DumpCache(bool ldElemCache)
+    {
+        PolymorphicInlineCache * cache = ldElemCache ? GetLdElemInlineCache() : GetStElemInlineCache();
+        Output::Print(_u("PropertyCache HitRate: %i; types: "), this->hitRate);
+        for (uint i = 0; i < cache->GetSize(); ++i)
         {
-            PolymorphicInlineCache * cache = ldElemCache ? GetLdElemInlineCache() : GetStElemInlineCache();
-            Output::Print(_u("PropertyCache HitRate: %i; types: "), this->hitRate);
-            for (uint i = 0; i < cache->GetSize(); ++i)
-            {
-                Output::Print(_u("%p,"), cache->GetInlineCaches()[i].GetType());
-            }
-            Output::Print(_u("\n"));
+            Output::Print(_u("%p,"), cache->GetInlineCaches()[i].GetType());
         }
+        Output::Print(_u("\n"));
+    }
 #endif
 #if ENABLE_TTD
-        //Get the associated property id for this string if there is on (e.g. it is a propertystring otherwise return Js::PropertyIds::_none)
-        virtual Js::PropertyId TryGetAssociatedPropertyId() const override { return this->propertyRecord->GetPropertyId(); }
+    //Get the associated property id for this string if there is on (e.g. it is a propertystring otherwise return Js::PropertyIds::_none)
+    virtual Js::PropertyId TryGetAssociatedPropertyId() const override { return this->propertyRecord->GetPropertyId(); }
 #endif
 
-    public:
-        virtual VTableValue DummyVirtualFunctionToHinderLinkerICF()
-        {
-            return VTableValue::VtablePropertyString;
-        }
-    };
+public:
+    virtual VTableValue DummyVirtualFunctionToHinderLinkerICF()
+    {
+        return VTableValue::VtablePropertyString;
+    }
+};
+
+
+template <typename T> inline
```

Comment:
why does this need to be templated? is it not guaranteed to always be a Var?

Reply:
You're right. The idea is that it might be a RecyclableObject, in which case we don't need check for RecyclableObject in the Is method. I'll add a comment.

Reply:
thanks!